src/protocols/screencopy.rs

```diff
@@ -275,44 +287,70 @@ where
 
         let (buffer, with_damage) = match request {
             zwlr_screencopy_frame_v1::Request::Copy { buffer } => (buffer, false),
-            // zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
+            zwlr_screencopy_frame_v1::Request::CopyWithDamage { buffer } => (buffer, true),
             _ => unreachable!(),
         };
 
-        if !shm::with_buffer_contents(&buffer, |_buf, shm_len, buffer_data| {
+        let buffer = if let Ok(dmabuf) = dmabuf::get_dmabuf(&buffer) {
+            if !(dmabuf.format().code == Fourcc::Argb8888
+                && dmabuf.size().w == info.buffer_size.w
+                && dmabuf.size().h == info.buffer_size.h)
+            {
+                frame.post_error(
+                    zwlr_screencopy_frame_v1::Error::InvalidBuffer,
+                    "invalid buffer",
+                );
+                return;
+            }
+            ScreencopyBuffer::Dmabuf(dmabuf.clone())
+        } else if shm::with_buffer_contents(&buffer, |_buf, shm_len, buffer_data| {
             buffer_data.format == wl_shm::Format::Argb8888
                 && buffer_data.stride == info.buffer_size.w * 4
                 && buffer_data.height == info.buffer_size.h
                 && shm_len as i32 == buffer_data.stride * buffer_data.height
         })
         .unwrap_or(false)
         {
+            ScreencopyBuffer::Shm(buffer)
+        } else {
             frame.post_error(
                 zwlr_screencopy_frame_v1::Error::InvalidBuffer,
                 "invalid buffer",
             );
             return;
-        }
+        };
 
         copied.store(true, Ordering::SeqCst);
 
-        state.frame(Screencopy {
-            with_damage,
+        let damages = state.frame(Screencopy {
             buffer,
             frame: frame.clone(),
             info: info.clone(),
             submitted: false,
         });
+
+        if with_damage {
+            if let Some(damages) = damages {
+                for Rectangle { loc, size } in damages {
+                    frame.damage(loc.x as u32, loc.y as u32, size.w as u32, size.h as u32);
```

Comment:
This should be sent before ready:

> This event is sent right before the ready event when copy_with_damage is requested.

And ready is currently sent in state.frame() above.